### PR TITLE
Foundry Invariant test Sample

### PR DESCRIPTION
**Pull request by prat-gpt**
Add an invariant test for adding and removing quote tokens in a fuzzed manner with a set of invariants that are checked after evert action.

This invariant test is based on test in following maple repository and foundry examples:
https://github.com/maple-labs/revenue-distribution-token/blob/add-forge-invariants/contracts/test/Invariants.t.sol

Foundry references:
https://github.com/foundry-rs/book/issues/497
https://github.com/foundry-rs/foundry/tree/master/testdata%2Ffuzz%2Finvariant

### CODE REVIEW

File: tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+pragma solidity 0.8.14;
+
+import "ds-test/test.sol";
+import "forge-std/console.sol";
+
+import { ERC20Pool }        from 'src/ERC20Pool.sol';
+import { ERC20PoolFactory } from 'src/ERC20PoolFactory.sol';
+import { Token }            from '../../utils/Tokens.sol';
+
+uint constant MIN_BUCKET_INDEX = 2570;
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol]: let's stick with `uint256` for consistency. Can we also make the range of buckets (MIN / MAX index) fuzzable too?

File: tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    // Mainnet ajna address
+    address internal _ajna = 0x9a96ec9B57Fb64FbC60B423d1f4da7691Bd35079;
+    ERC20Pool internal _pool;
+    Token internal _collateral;
+    Token internal _quote;
+    ERC20PoolFactory internal _poolFactory;
+
+    function setUp() public virtual {
+        _collateral  = new Token("Collateral", "C");
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol]: can we fuzz collateral and quote precision?

File: tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        _poolFactory = new ERC20PoolFactory(_ajna);
+        _pool        = ERC20Pool(_poolFactory.deployPool(address(_collateral), address(_quote), 0.05 * 10**18));
+        _invariantLenderManager = new InvariantLenderManager(address(_pool), address(_quote));
+
+        // create first lender
+        _invariantLenderManager.createLender();
+    }
+
+    // include only invariantLenderManager contract for invariant testing
+    function targetContracts() public view returns (address[] memory) {
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol]: is this only a view function for info or should it be used in the invariant tests somehow?

**Review by grandizzy — COMMENTED** [review]
would be nice to have a `make test-invariants` target that would execute / match tests or classes with `Invariant`

File: tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol (hunk 3 above)
Comment (prat-gpt) [reply tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol]: Foundry looks for this function to target the contracts whose methods should be called while invariant testing. We can also exclude contracts.

File: tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol (hunk 2 above)
Comment (prat-gpt) [reply tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol]: Yes, right!. Will work on that once basic setup is running properly :).

File: tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol (hunk 3 above)
Comment (grandizzy) [reply tests/forge/ERC20Pool/invariants/ERC20PoolQuoteTokenInvariant.t.sol]: ah, nice, TIL something, thank you!

**Comment by grandizzy** [comment]
merged another PR, closing this one

**Pull request closed without merging**